src/main.rs

```diff
@@ -169,6 +175,88 @@ fn resolve_canister_id(
     None
 }
 
+fn decode_hash_tree(
+    name: &str,
+    value: Option<String>,
+    logger: &slog::Logger,
+) -> Result<Vec<u8>, ()> {
+    match value {
+        Some(tree) => base64::decode(tree).map_err(|e| {
+            slog::warn!(logger, "Unable to decode {} from base64: {}", name, e);
+        }),
+        _ => Err(()),
+    }
+}
+
+struct HeadersData {
+    certificate: Option<Result<Vec<u8>, ()>>,
+    tree: Option<Result<Vec<u8>, ()>>,
+    encoding: Option<String>,
+}
+
+fn extract_headers_data(headers: &[HeaderField], logger: &slog::Logger) -> HeadersData {
+    let mut headers_data = HeadersData {
+        certificate: None,
+        tree: None,
+        encoding: None,
+    };
+
+    for HeaderField(name, value) in headers {
+        if name.eq_ignore_ascii_case("IC-CERTIFICATE") {
+            for field in value.split(',') {
+                if let Some((_, name, b64_value)) = regex_captures!("^(.*)=:(.*):$", field.trim()) {
+                    slog::trace!(logger, ">> certificate {}: {}", name, b64_value);
```

Comment:
Are there limits on the size of these traced values ?

Reply:
yes, this is the maximum size of base64-encoded IC-certificate. In reality it is no more than ~0-500kb